File: python/sort/insertion_sort.py
"""
Recursive insertion sort implementation in Python 3
https://en.wikipedia.org/wiki/Insertion_sort
"""


def insertion_sort(lis):
    """
    Implementation of recursive insertion sort.
    :param lis: list to be sorted
    """
    for i in range(len(lis)):
        j = i
        while j > 0 and lis[j-1] > lis[j]:
            lis[j], lis[j-1] = lis[j-1], lis[j]
            j -= 1
    return lis


def main():
    lis1 = [4, 1, 2, 3, 9]
    lis2 = [1]
    lis3 = [2, 2, 1, -1, 0, 4, 5, 2]
    lis4 = []

    lis1 = insertion_sort(lis1)
    assert lis1 == [1, 2, 3, 4, 9]

    lis2 = insertion_sort(lis2)
    assert lis2 == [1]

    lis3 = insertion_sort(lis3)
    assert lis3 == [-1, 0, 1, 2, 2, 2, 4, 5]

    lis4 = insertion_sort(lis4)
    assert lis4 == []

if __name__ == '__main__':
    main()

File: cpp/sort/insertion_sort.cpp
/*
	Implementation of Insertion Sort in c++(gcc 4.9.2)
	https://en.wikipedia.org/wiki/Insertion_sort
*/

#include <iostream>
using namespace std;
int array[10000];

void insertion_sort(int n)
{
	int i,j,key;
	for(i=1;i<n;i++)
	{
		key=array[i];
		j=i-1;
		while(j>=0 && key<array[j])
		{
			array[j+1]=array[j];
			j--;
		}
		array[j+1]=key;
	}
}

int main()
{
	int n,i;
	cout<<"Input length of array- ";
	cin>>n;
	for(i=0;i<n;i++)
		cin>>array[i];
	insertion_sort(n);
	cout<<"Sorted Array is-\n";
	for(i=0;i<n;i++)
		cout<<array[i]<<" ";
	return 0;
}